lib/actions/usbsdfuTargetActions.js

```diff
@@ -268,103 +287,196 @@ function calculateSHA512Hash(image) {
-                logger.info('DFU completed successfully!');
                 dispatch(targetActions.writeProgressFinishedAction());
-                dispatch({ type: 'DEVICE_DESELECTED' });
-            })
-            .catch(error => {
-                logger.error(`DFU failed. Reason: ${error.message || error}`);
-            })
-            .then(() => {
-                port.close();
-                logger.info(`Port is closed: ${port.path}`);
+                dispatch(canWrite());
             });
+    }
+
+    const dfuImages = inputDfuImages;
+    let image;
+    let serialNumber;
+
+    return async (dispatch, getState) => Promise.resolve()
+        // When dfu starts
+        .then(async () => {
+            serialNumber = getState().app.target.serialNumber;
+            image = dfuImages.first();
+            const packet = await initPacket.createInitPacketUint8Array(image.initPacket);
+            const firmwareUpdates = new dfujs.DfuUpdates([{
+                initPacket: packet,
+                firmwareImage: image.firmwareImage,
+            }]);
+            const serialUsbTransport = new dfujs.DfuTransportUsbSerial(serialNumber);
+            const dfu = new dfujs.DfuOperation(firmwareUpdates, serialUsbTransport);
+
+            dispatch(targetActions.targetWritableKnownAction(false));
+            dispatch(targetActions.writeProgressStartAction());
+            logger.info('Performing DFU. This may take a few seconds');
+
+            return dfu.start(true);
+        })
+        // Catch error during dfu
+        .catch(error => {
+            logger.error(`DFU failed. Reason: ${error.message || error}`);
+            throw error;
+        })
+        // Wait for the device to reattach
+        .then(() => {
+            logger.info(`DFU for ${dfuImages.first().name} completed successfully!`);
+            logger.info(`${dfuImages.size - 1} dfu pakcage(s) left.`);
+            logger.info('Waiting for device');
+            if (image.name.includes(RegionName.BOOTLOADER) ||
+                image.name.includes(RegionName.SOFTDEVICE)) {
+                return waitForDevice(serialNumber, 5000, ['serialport']);
+            } else if (image.name.includes(RegionName.APPLICATION)) {
+                return waitForDevice(serialNumber, 5000, ['nordicusb']);
+            }
+            throw new Error('Device type is unknown when waiting for device.');
+        })
+        // Select and wait for device to open
+        .then(device => dispatch(selectAndWaitForDeviceOpen(device)))
+        // Remove the current dfu image and continue with the next
+        .then(() => {
+            dispatch(operateDFU(dfuImages.delete(0)));
+        })
+        // Catch error when reopening device
+        .catch(error => {
+            logger.error(`Reopen device failed. Reason: ${error.message || error}`);
+        });
+}
+
+// Write files to target device
+export function write() {
+    return async (dispatch, getState) => {
```

Comment:
I think `async` is not needed unless there's an `await`.

Reply:
Yes, I have moved all the await out. Will remove the async.